src/mongo/client/gridfs.cpp

```diff
@@ -321,14 +321,21 @@ namespace mongo {
             invariant( _pendingDataSize <= _chunkSize );
             if (_pendingDataSize == _chunkSize) {
                 _appendPendingData();
-                _appendChunk( data + size, length - size, false );
+                const char* const end = data + length;
+                data = _appendChunk( data + size, length - size, false );
+                if (data != end) {
+                    invariant(data < end);
+                    size_t nsize = static_cast<size_t>(end - data);
+                    memcpy( _pendingData.get() + _pendingDataSize, data, nsize );
+                    _pendingDataSize += nsize;
+                }
             }
         }
         else {
             const char* const end = data + length;
             // split data in _chunkSize blocks, and store as pending the last block if
             // necessary
-            data = _appendChunk( data, length, false );
+            data = _appendChunk( data, length, false);
```

Comment:
Unrelated? I know the formatting is inconsistent in this file, but I think you should restore the space you removed here.